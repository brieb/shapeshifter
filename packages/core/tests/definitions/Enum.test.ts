import { options } from '../mocks';
import EnumDefinition from '../../src/definitions/Enum';

describe('definitions/Enum', () => {
  it('errors if `valueType` is missing', () => {
    expect(() => new EnumDefinition(options, 'foo', {})).toThrow(
      'Invalid EnumDefinition field "valueType". Field is required and must be defined.',
    );
  });

  it('errors if `valueType` is empty', () => {
    expect(
      () => new EnumDefinition(options, 'foo', { valueType: '' }),
    ).toThrowErrorMatchingSnapshot();
  });

  it('errors if `values` is not an array', () => {
    expect(
      // @ts-ignore
      () => new EnumDefinition(options, 'foo', { valueType: 'string', values: 'string' }),
    ).toThrow('Invalid EnumDefinition field "values". Must be an array.');
  });

  it('errors if `values` is empty', () => {
    expect(
      () => new EnumDefinition(options, 'foo', { valueType: 'string', values: [] }),
    ).toThrow('Invalid EnumDefinition field "values". Array cannot be empty.');
  });

  describe('errors if `values` do not match the type in `valueType`', () => {
    it('for strings', () => {
      expect(
        () => new EnumDefinition(options, 'foo', { valueType: 'string', values: ['foo', 123] }),
      ).toThrow('Enum values do not match the defined value type.');
    });

    it('for booleans', () => {
      expect(
        () => new EnumDefinition(options, 'foo', { valueType: 'bool', values: [true, 123] }),
      ).toThrow('Enum values do not match the defined value type.');
    });

    it('for numbers', () => {
      expect(
        () => new EnumDefinition(options, 'foo', { valueType: 'number', values: [123, 'foo'] }),
      ).toThrow('Enum values do not match the defined value type.');
    });
  });

  describe('does not error if `values` match the type in `valueType`', () => {
    it('for strings', () => {
      expect(
        () => new EnumDefinition(options, 'foo', { valueType: 'string', values: ['foo', 'bar'] }),
      ).not.toThrow();
    });

    it('for booleans', () => {
      expect(
        () => new EnumDefinition(options, 'foo', { valueType: 'bool', values: [true, false] }),
      ).not.toThrow();
    });

    it('for numbers', () => {
      expect(
        () => new EnumDefinition(options, 'foo', { valueType: 'number', values: [123, 456.789] }),
      ).not.toThrow();
    });
  });

  it('does not error if `constant` is true', () => {
    expect(
      () =>
        new EnumDefinition(options, 'foo', {
          valueType: 'number',
          values: ['foo', 'bar'],
          constant: true,
        }),
    ).not.toThrow();
  });
});
