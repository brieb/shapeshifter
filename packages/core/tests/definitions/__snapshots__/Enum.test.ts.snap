// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`definitions/Enum errors if \`valueType\` is empty 1`] = `
"Invalid EnumDefinition field \\"valueType\\". Type must be one of: string, shape. Received string with the following invalidations:
 - String cannot be empty."
`;
